README.md
# flutter_placeholder_textlines

A simple plugin to generate placeholder lines that emulates text in a UI, useful for displaying placeholder content while loading or empty items

## Example of Usage

```dart
Widget _buildCustomColorSizePlaceholder() {
    return Container(
      width: 200,
      child: PlaceholderLines(
        count: 4,
        align: TextAlign.center,
        lineHeight: 8,
        color: Colors.blueAccent,
      ),
    );
  }

  Widget _buildSimpleTextPlaceholder(TextAlign align) {
    return Container(
      width: 300,
      child: PlaceholderLines(
        count: 3,
        align: align,
      ),
    );
  }

  Widget _buildSubtitle(String title) {
    return Container(
      padding: EdgeInsets.only(
        top: 24,
        bottom: 16,
      ),
      child: Text(
        title,
        style: TextStyle(fontSize: 20),
      ),
    );
  }

  _buildAnimated() {
    return Container(
      width: 300,
      child: PlaceholderLines(
        count: 3,
        animate: true,
        color: Colors.purple,
      ),
    );
  }

  _buildCardExample() {
    return Material(
      borderRadius: BorderRadius.circular(10),
      elevation: 9,
      child: Container(
        padding: EdgeInsets.symmetric(vertical: 16, horizontal: 20),
        width: 300,
        child: Row(
          children: <Widget>[
            Container(
              margin: EdgeInsets.only(right: 16),
              width: 70,
              height: 70,
              decoration: BoxDecoration(color: Colors.grey.withOpacity(.6), ),
              child: Center(child: Icon(Icons.photo_size_select_actual, color: Colors.white, size: 38,),),
            ),
            Expanded(
              child: PlaceholderLines(
                count: 3,
              ),
            ),
          ],
        ),
      ),
    );
  }
  ```

  ## API Options

  ```dart
  /// Defines how many lines to build
  final int count;

  /// Defines `color` to be used as base when building each line
  final Color color;

  /// Defines `alignment` for all the lines generated
  final TextAlign align;

  /// Defines the `min` value for the `opacity` that a line could have, since color opacity it's generated randomly ,
  /// this prevents the value from going to low (making it invisible)
  /// defaults to [0.4]
  final double minOpacity;

  /// Defines the `max` value for the `opacity` that a line could have, since color opacity it's generated randomly ,
  /// defaults to [0.94]
  final double maxOpacity;

  /// Defines the `max` `width` that a line could have, since `width` it's generated randomly
  /// defaults to [.95]
  final double maxWidth;

  /// Defines the `min` `width` that a line could have, since `width` it's generated randomly
  /// defaults to [.72]
  final double minWidth;

  /// Defines the line `height`
  /// defaults to [12]
  final double lineHeight;

  /// if [true], plays a nice animation of an overlay
  final bool animate;

  /// Use a [customAnimationOverlay] to display instead of the difault one
  final Widget customAnimationOverlay;

  /// Set a custom [animationOverlayColor]
  final Color animationOverlayColor;

  /// If [true] , this will cause the lines to be rebuild with different widths (randomly generated)
  /// every time any parent widget rebuilds it's state
  /// defaults to [false]
  final bool rebuildOnStateChange;
  ```

  Also please see example at: https://github.com/victorevox/flutter_placeholder_textlines/tree/master/example


  ## Screenshots

  <img src="https://raw.githubusercontent.com/victorevox/flutter_placeholder_textlines/master/example/Screenshot_1.png" align = "left" height = "350" alt="examples">

  <img src="https://raw.githubusercontent.com/victorevox/flutter_placeholder_textlines/master/example/Screenshot_1.png" align = "left" height = "350" alt="examples 2">

  <img src="https://raw.githubusercontent.com/victorevox/flutter_placeholder_textlines/master/example/example.gif" height = "350" alt="examples gif">


  ## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.

CHANGELOG.md

## [1.1.2] - 18/08/21
* Some fix
 
## [1.1.1] - 09/08/21
* Some small improvements
  
## [1.1.0] - 09/08/21
* Version change
## [1.0.5] - 14/06/20

* Null safety

## [1.0.4] - 25/02/20

* Improve performance for not animated widgets.

## [1.0.3] - 30/01/20
* Fixed issue where context was expected but null was found
* Fixed issue where each line was rebuild with a different width every time a parent widget rebuild it's state, added a property to enable this behaviour if wanted (```rebuildOnStateChange```)
* Updated README.md
  
## [1.0.2] - 26/01/20
* Dispose animation controller
  
## [1.0.1] - 26/01/20
* Fix unexported widget

## [1.0.0] - 25/01/20

* Initial release


test/flutter_placeholder_textlines_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_placeholder_textlines/flutter_placeholder_textlines.dart';

void main() {
  test('adds one to input values', () {
    
  });
}


lib/placeholder_lines.dart
import 'dart:math';

import 'package:flutter/widgets.dart';

class PlaceholderLines extends StatefulWidget {
  /// Defines how many lines to build
  final int count;

  /// Defines `color` to be used as base when building each line
  final Color color;

  /// Defines `alignment` for all the lines generated
  final TextAlign align;

  /// Defines the `min` value for the `opacity` that a line could have, since color opacity it's generated randomly ,
  /// this prevents the value from going to low (making it invisible)
  /// defaults to [0.4]
  final double minOpacity;

  /// Defines the `max` value for the `opacity` that a line could have, since color opacity it's generated randomly ,
  /// defaults to [0.94]
  final double maxOpacity;

  /// Defines the `max` `width` that a line could have, since `width` it's generated randomly
  /// defaults to [.95]
  final double maxWidth;

  /// Defines the `min` `width` that a line could have, since `width` it's generated randomly
  /// defaults to [.72]
  final double minWidth;

  /// Defines the line `height`
  /// defaults to [12]
  final double lineHeight;

  /// if [true], plays a nice animation of an overlay
  final bool animate;

  /// Use a [customAnimationOverlay] to display instead of the difault one
  final Widget? customAnimationOverlay;

  /// Set a custom [animationOverlayColor]
  final Color? animationOverlayColor;

  /// If [true] , this will cause the lines to be rebuild with different widths (randomly generated)
  /// every time any parent widget rebuilds it's state
  /// defaults to [false]
  final bool rebuildOnStateChange;

  const PlaceholderLines({
    Key? key,
    required this.count,
    this.align = TextAlign.left,
    this.color = const Color(0xFFDEDEDE),
    this.minOpacity = 0.4,
    this.maxOpacity = 0.94,
    this.maxWidth = .95,
    this.minWidth = .72,
    this.lineHeight = 12,
    this.animate = false,
    this.customAnimationOverlay,
    this.animationOverlayColor,
    this.rebuildOnStateChange = false,
  })  : assert(count is int),
        assert(minOpacity <= 1 && maxOpacity <= 1),
        assert(minWidth <= 1 && maxWidth <= 1),
        assert(minOpacity is double),
        assert(maxOpacity is double),
        assert(minOpacity <= maxOpacity),
        super(key: key);

  @override
  _PlaceholderLinesState createState() => _PlaceholderLinesState();
}

class _PlaceholderLinesState extends State<PlaceholderLines>
    with SingleTickerProviderStateMixin {
  late AnimationController _animationController;
  Animation<RelativeRect>? _animation;
  late Map<int, double> _seeds;

  double get _randomSeed => Random().nextDouble();

  @override
  void didUpdateWidget(PlaceholderLines oldWidget) {
    if (oldWidget.animate != widget.animate) {
      if (widget.animate) {
        // _setupAnimation();
        _animationController.forward();
      } else {
        // if(_animationController != null) {
        //   _animationController.dispose();
        //   _animationController = null;
        // }
        _animationController.stop();
      }
    }
    super.didUpdateWidget(oldWidget);
  }

  @override
  void initState() {
    _seeds = List.filled(widget.count, 0).asMap().map((index, _) {
      return MapEntry(index, _randomSeed);
    });

    _animationController = AnimationController(
      vsync: this,
      duration: Duration(seconds: 1),
    );

    WidgetsFlutterBinding.ensureInitialized();
    WidgetsBinding.instance!.addPostFrameCallback(_setupAnimation);
    // if(widget.animate) {
    // }
    super.initState();
  }

  void _setupAnimation([Duration? _]) {
    if(!mounted) return;
    
    final RenderBox? renderO = context.findRenderObject() as RenderBox;

    if (renderO == null) {
      throw Exception(
          "RenderBox not found -- could not calculate BoxConstraints for "
          "placeholder_lines");
    }

    final BoxConstraints constraints = renderO.constraints;
    final double maxWidth = _getMaxConstrainedWidth(constraints);
    final CurvedAnimation curvedAnimation = CurvedAnimation(
      curve: Curves.easeIn,
      reverseCurve: Curves.easeInOut,
      parent: _animationController,
    );
    _animation = RelativeRectTween(
      begin: RelativeRect.fromLTRB(0, 0, maxWidth, 0),
      end: RelativeRect.fromLTRB(maxWidth, 0, -maxWidth, 0),
    ).animate(curvedAnimation)
      ..addStatusListener(
        (status) {
          if (!widget.animate) return;
          if (status == AnimationStatus.completed) {
            _animationController.reverse();
          } else if (status == AnimationStatus.dismissed) {
            _animationController.forward();
          }
        },
      );
    setState(() {
      _animationController.forward();
    });
  }

  @override
  Widget build(BuildContext context) {
    CrossAxisAlignment _alignment = this.widget.align == TextAlign.left
        ? CrossAxisAlignment.start
        : this.widget.align == TextAlign.right
            ? CrossAxisAlignment.end
            : this.widget.align == TextAlign.center
                ? CrossAxisAlignment.center
                : CrossAxisAlignment.baseline;
    return LayoutBuilder(
      builder: (context, constraints) {
        return Container(
          width: constraints.maxWidth,
          child: Column(
            mainAxisSize: MainAxisSize.min,
            crossAxisAlignment: _alignment,
            children: _buildLines(constraints),
          ),
        );
      },
    );
  }

  List<Widget> _buildLines(BoxConstraints constraints) {
    List<Widget> list = [];
    for (var i = 0; i < widget.count; i++) {
      double _random = widget.rebuildOnStateChange ? _randomSeed : _seeds[i]!;
      double _opacity = (widget.maxOpacity -
              ((widget.maxOpacity - widget.minOpacity) * _random))
          .abs();
      // double realMaxWidth = constraints.maxWidth;
      double constrainedMaxWidth = _getMaxConstrainedWidth(constraints);
      double constrainedMinWidth = _getMinConstrainedWidth(constraints);
      double _width = (constrainedMaxWidth -
              ((constrainedMaxWidth - constrainedMinWidth) * _random))
          .abs();
      final Widget staticWidget = Container(
        height: widget.lineHeight,
        width: _width,
        decoration: BoxDecoration(
          color: widget.color.withOpacity(_opacity),
        ),
      );
      list.add(
        AnimatedSwitcher(duration: Duration(milliseconds: 400),child: _animation != null && widget.animate == true
            ? AnimatedBuilder(
                animation: _animation!,
                child: staticWidget,
                builder: (context, child) {
                  return Container(
                    // decoration: BoxDecoration(color: Colors.purple),
                    child: ClipRect(
                      child: Stack(
                        clipBehavior: Clip.antiAlias,
                        children: <Widget>[
                          if (child != null) child,
                          Transform.translate(
                            offset: Offset(_animation!.value.left, 0),
                            child: widget.customAnimationOverlay ??
                                Container(
                                  height: widget.lineHeight,
                                  width: _width / 3,
                                  decoration: BoxDecoration(boxShadow: [
                                    BoxShadow(
                                        blurRadius: 18,
                                        color: widget.animationOverlayColor ??
                                            Color(0xFFFFFFFF),
                                        offset: Offset(4, 3)),
                                    BoxShadow(
                                        blurRadius: 28,
                                        color: widget.animationOverlayColor ??
                                            Color(0xFFFFFFFF),
                                        offset: Offset(1, 3)),
                                  ]),
                                ),
                          )
                        ],
                      ),
                    ),
                  );
                },)
            : staticWidget,),
      );
      if (i < widget.count - 1) {
        list.add(SizedBox(
          height: widget.lineHeight,
        ));
      }
    }
    return list;
  }

  double _getMaxConstrainedWidth(BoxConstraints constraints) {
    double realMaxWidth = constraints.maxWidth;
    return realMaxWidth * widget.maxWidth;
  }

  double _getMinConstrainedWidth(BoxConstraints constraints) {
    double realMaxWidth = constraints.maxWidth;
    return realMaxWidth * widget.minWidth;
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }
}

example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_placeholder_textlines/placeholder_lines.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    String title = "flutter_placeholder_textlines Demo";
    return MaterialApp(
      title: title,
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: title),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SingleChildScrollView(
        child: Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              Column(
                children: <Widget>[
                  _buildSubtitle("Centered"),
                  _buildSimpleTextPlaceholder(TextAlign.center),
                  _buildSubtitle("Left"),
                  _buildSimpleTextPlaceholder(TextAlign.left),
                  _buildSubtitle("Right"),
                  _buildSimpleTextPlaceholder(TextAlign.right),
                  _buildSubtitle("Change color/size"),
                  _buildCustomColorSizePlaceholder(),
                  _buildSubtitle("Animated"),
                  _buildAnimated(),
                  _buildSubtitle("Card Example"),
                  _buildCardExample(),
                  SizedBox(
                    height: 20,
                  ),
                ],
              ),
            ],
          ),
        ),
      ),
    );
  }

  Widget _buildCustomColorSizePlaceholder() {
    return Container(
      width: 200,
      child: PlaceholderLines(
        count: 4,
        align: TextAlign.center,
        lineHeight: 8,
        color: Colors.blueAccent,
      ),
    );
  }

  Widget _buildSimpleTextPlaceholder(TextAlign align) {
    return Container(
      width: 300,
      child: PlaceholderLines(
        count: 3,
        align: align,
      ),
    );
  }

  Widget _buildSubtitle(String title) {
    return Container(
      padding: EdgeInsets.only(
        top: 24,
        bottom: 16,
      ),
      child: Text(
        title,
        style: TextStyle(fontSize: 20),
      ),
    );
  }

  _buildAnimated() {
    return AnimatedWrapper();
  }

  _buildCardExample() {
    return Material(
      borderRadius: BorderRadius.circular(10),
      elevation: 9,
      child: Container(
        padding: EdgeInsets.symmetric(vertical: 16, horizontal: 20),
        width: 300,
        child: Row(
          children: <Widget>[
            Container(
              margin: EdgeInsets.only(right: 16),
              width: 70,
              height: 70,
              decoration: BoxDecoration(
                color: Colors.grey.withOpacity(.6),
              ),
              child: Center(
                child: Icon(
                  Icons.photo_size_select_actual,
                  color: Colors.white,
                  size: 38,
                ),
              ),
            ),
            Expanded(
              child: PlaceholderLines(
                count: 3,
              ),
            ),
          ],
        ),
      ),
    );
  }
}

class AnimatedWrapper extends StatefulWidget {
  const AnimatedWrapper({
    Key key,
  }) : super(key: key);

  @override
  _AnimatedWrapperState createState() => _AnimatedWrapperState();
}

class _AnimatedWrapperState extends State<AnimatedWrapper> {
  bool _animated = true;
  @override
  void initState() {
    super.initState();
    _changeState();
  }

  _changeState() {
    setState(() {
      _animated = !_animated;
    });
    Future.delayed(Duration(seconds: 5)).then((_) => _changeState());
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      width: 300,
      child: PlaceholderLines(
        count: 3,
        animate: _animated,
        color: Colors.purple,
      ),
    );
  }
}
